File: cpp/2loops.cpp
#include <chrono>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <vector>

using namespace std;
using reloj = chrono::high_resolution_clock;

static void llenar_datos(vector<double>& A, vector<double>& x, int n) {
    for (int i = 0; i < n; i++) {
        x[i] = 1.0 + (i % 7);
        for (int j = 0; j < n; j++) {
            A[(size_t)i * n + j] = (double)((i + j) % 100) / 25.0;
        }
    }
}

static double ejecutar_orden_ij(const vector<double>& A, const vector<double>& x, vector<double>& y, int n) {
    fill(y.begin(), y.end(), 0.0);

    auto inicio = reloj::now();
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            y[i] += A[(size_t)i * n + j] * x[j];
        }
    }
    auto fin = reloj::now();

    return chrono::duration<double, milli>(fin - inicio).count();
}

static double ejecutar_orden_ji(const vector<double>& A, const vector<double>& x, vector<double>& y, int n) {
    fill(y.begin(), y.end(), 0.0);

    auto inicio = reloj::now();
    for (int j = 0; j < n; j++) {
        for (int i = 0; i < n; i++) {
            y[i] += A[(size_t)i * n + j] * x[j];
        }
    }
    auto fin = reloj::now();

    return chrono::duration<double, milli>(fin - inicio).count();
}

static double checksum(const vector<double>& y) {
    double suma = 0.0;
    for (double valor : y) {
        suma += valor;
    }
    return suma;
}

int main(int argc, char* argv[]) {
    vector<int> tamanos;

    if (argc > 1) {
        for (int i = 1; i < argc; i++) {
            tamanos.push_back(atoi(argv[i]));
        }
    } else {
        tamanos = {256, 512, 1024};
    }

    cout << "Comparacion de dos bucles anidados para y = A*x" << '\n';
    cout << left << setw(10) << "N"
         << setw(18) << "orden i-j (ms)"
         << setw(18) << "orden j-i (ms)"
         << setw(14) << "razon"
         << "checksum" << '\n';

    for (int n : tamanos) {
        vector<double> A((size_t)n * n);
        vector<double> x(n);
        vector<double> y(n);

        llenar_datos(A, x, n);

        double tiempo_ij = ejecutar_orden_ij(A, x, y, n);
        double control = checksum(y);

        double tiempo_ji = ejecutar_orden_ji(A, x, y, n);
        control += checksum(y);

        cout << left << setw(10) << n
             << setw(18) << fixed << setprecision(3) << tiempo_ij
             << setw(18) << fixed << setprecision(3) << tiempo_ji
             << setw(14) << fixed << setprecision(2) << (tiempo_ji / tiempo_ij)
             << fixed << setprecision(2) << control << '\n';
    }

    return 0;
}

File: go/2loops.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func llenarDatos(A []float64, x []float64, n int) {
	for i := 0; i < n; i++ {
		x[i] = 1.0 + float64(i%7)
		for j := 0; j < n; j++ {
			A[i*n+j] = float64((i+j)%100) / 25.0
		}
	}
}

func ejecutarOrdenIJ(A []float64, x []float64, y []float64, n int) float64 {
	for i := range y {
		y[i] = 0.0
	}

	inicio := time.Now()
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			y[i] += A[i*n+j] * x[j]
		}
	}

	return time.Since(inicio).Seconds() * 1000.0
}

func ejecutarOrdenJI(A []float64, x []float64, y []float64, n int) float64 {
	for i := range y {
		y[i] = 0.0
	}

	inicio := time.Now()
	for j := 0; j < n; j++ {
		for i := 0; i < n; i++ {
			y[i] += A[i*n+j] * x[j]
		}
	}

	return time.Since(inicio).Seconds() * 1000.0
}

func checksum(y []float64) float64 {
	suma := 0.0
	for _, valor := range y {
		suma += valor
	}
	return suma
}

func main() {
	tamanos := []int{256, 512, 1024}

	if len(os.Args) > 1 {
		tamanos = nil
		for _, arg := range os.Args[1:] {
			n, err := strconv.Atoi(arg)
			if err == nil && n > 0 {
				tamanos = append(tamanos, n)
			}
		}
	}

	fmt.Println("Comparacion de dos bucles anidados para y = A*x")
	fmt.Printf("%-10s%-18s%-18s%-14s%s\n", "N", "orden i-j (ms)", "orden j-i (ms)", "razon", "checksum")

	for _, n := range tamanos {
		A := make([]float64, n*n)
		x := make([]float64, n)
		y := make([]float64, n)

		llenarDatos(A, x, n)

		tiempoIJ := ejecutarOrdenIJ(A, x, y, n)
		control := checksum(y)

		tiempoJI := ejecutarOrdenJI(A, x, y, n)
		control += checksum(y)

		fmt.Printf("%-10d%-18.3f%-18.3f%-14.2f%.2f\n", n, tiempoIJ, tiempoJI, tiempoJI/tiempoIJ, control)
	}
}
